### PR TITLE
CT: Add interval solver

### PR DESCRIPTION
**Pull request by facuMH**
This PR adds an interval solver. This tool can be used to describe a set of intervals and get a random number uniformly from all the possible values considering all the intervals. 

This PR contributes to #459

### CODE REVIEW

File: go/ct/gen/interval_solver.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+func (s *IntervalSolver[T]) AddUpperBoundary(value T) {
+	if len(s.intervals) == 0 {
+		return
+	}
+	max := s.intervals[len(s.intervals)-1].high
+	s.Exclude(value+1, max)
+}
+
+func (s *IntervalSolver[T]) IsSatisfiable() bool {
+	return len(s.intervals) > 0
```
Comment (LuisPH3) [inline go/ct/gen/interval_solver.go:80]: Because the empty interval exists, the solver may not be satisfiable if initialized with only empty intervals.

File: go/ct/gen/interval_solver.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	if len(s.intervals) == 0 {
+		return 0, ErrUnsatisfiable
+	}
+
+	domainSize := uint64(0)
+	for _, interval := range s.intervals {
+		domainSize += uint64(interval.high - interval.low + 1)
+	}
+
+	if domainSize == 0 { // the domain is the full uint64 domain
```
Comment (LuisPH3) [inline go/ct/gen/interval_solver.go:93]: If domain size is less than 0, because of invalid/empty intervals, there should be an error

File: go/ct/gen/interval_solver_test.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+	"pgregory.net/rand"
+)
+
+func TestIntervalSolver_CanFormulateConstraints(t *testing.T) {
+	tests := map[string]struct {
+		setup func(*IntervalSolver[int32])
+		want  string
+	}{
+		"empty": {
```
Comment (LuisPH3) [inline go/ct/gen/interval_solver_test.go:14]: missleading, rename to default.

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
It is really ingenuous. 
Please solve the issue when solver is initialized to the empty interval.

File: go/ct/gen/interval_solver.go (hunk 1 above)
Comment (facuMH) [reply go/ct/gen/interval_solver.go:84]: great observations !! tests and checks added for empty intervals aka `[a..b]` where `a>b`

File: go/ct/gen/interval_solver.go (hunk 2 above)
Comment (facuMH) [reply go/ct/gen/interval_solver.go:100]: because `domainSize` is a `uint64` it can not be negative, it would be an overflow, which could easily be confused with a big interval, for this a check was added to verify that the interval is not empty, aka `interval.low<interval.high`

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/gen/interval_solver.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/constraints"
+	"pgregory.net/rand"
+)
+
+type IntervalSolver[T constraints.Integer] struct {
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver.go:11]: Need documentation!

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/gen/interval_solver.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+func (s *IntervalSolver[T]) Generate(rnd *rand.Rand) (T, error) {
+	if s.intervals == nil || len(s.intervals) == 0 {
+		return 0, ErrUnsatisfiable
+	}
+
+	domainSize := uint64(0)
+	for _, interval := range s.intervals {
+		if interval.isEmpty() {
+			return 0, ErrUnsatisfiable
+		}
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver.go:96]: This should not happen. There should be an invariant that makes sure that there is no empty interval in the interval list.

File: go/ct/gen/interval_solver.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	clauses := []string{}
+	for _, interval := range s.intervals {
+		clauses = append(clauses, fmt.Sprintf("[%d..%d]", interval.low, interval.high))
+	}
+	return fmt.Sprintf("X ∈ %s", strings.Join(clauses, " ∪ "))
+}
+
+func (s *IntervalSolver[T]) Clone() *IntervalSolver[T] {
+	return &IntervalSolver[T]{intervals: append([]interval[T]{}, s.intervals...)}
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver.go:126]: nit: use `slices.Clone(s.intervals)`

File: go/ct/gen/interval_solver.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+func (s *IntervalSolver[T]) Equals(other *IntervalSolver[T]) bool {
+	if len(s.intervals) != len(other.intervals) {
+		return false
+	}
+	for i, interval := range s.intervals {
+		if interval != other.intervals[i] {
+			return false
+		}
+	}
+	return true
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver.go:138]: Use `slices.Equal()`

Comment (HerbertJordan) [inline go/ct/gen/interval_solver.go:138]: This function is only true if the following invariants are guaranteed:
- the intervals are sorted
- there are no overlaps; in particular, there are no empty intervals
The last one should be the case, but in the print function you are checking for this empty intervals. So this should be ensured.

File: go/ct/gen/interval_solver_test.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+	const N = 20
+	// remove interval [c..d] from [a..b]
+	for a := 0; a < N; a++ {
+		for b := 0; b < N; b++ {
+			for c := 0; c < N; c++ {
+				for d := 0; d < N; d++ {
+					if a > b || c > d {
+						// we skip empty intervals
+						continue
+					}
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:124]: Remove this => the implementation must be able to handle empty intervals.

File: go/ct/gen/interval_solver_test.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+			}
+		}
+	}
+}
+
+func TestIntervalSolver_EmptyInterval(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 100)
+	_, err := solver.Generate(rand.New())
+	if err != ErrUnsatisfiable {
+		t.Errorf("failed to generate with empty interval: %v", err)
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:146]:
```suggestion
		t.Errorf("failed to generate point in empty interval: %v", err)
```

File: go/ct/gen/interval_solver_test.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+	}
+}
+
+func TestIntervalSolver_AddEmptyIntervalHasNoEffect(t *testing.T) {
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:150]:
```suggestion
func TestIntervalSolver_ExcludeEmptyIntervalHasNoEffect(t *testing.T) {
```

File: go/ct/gen/interval_solver_test.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	if solver1.Equals(solver2) {
+		t.Errorf("equals fails to report different instances")
+	}
+}
+
+func TestInterval_isEmpty(t *testing.T) {
+	tests := map[string]struct {
+		interval interval[int32]
+		want     bool
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:234]: nit: would be more readable if you use the fields
```
low int32
high int32
empty bool
```

File: go/ct/gen/interval_solver_test.go
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:1]: Please add tests for additional (edge) cases, like:
- intervals covering the full 64-bit range of uint64
- intervals on unsigned types and 0 as a lower end
- intervals on signed types with negative values

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/gen/interval_solver.go (hunk 5 above)
Comment (LuisPH3) [reply go/ct/gen/interval_solver.go:96]: The way I see it, if you have a constructor function with 2 integers, and they get misplaced, the object will be constructed in an insatisfacible state. Would it?

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
The empty interval issue is now solved!. 
I am concerned about over/under flow when moving boundaries. :(

File: go/ct/gen/interval_solver_test.go [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+			}
+		}
+	}
+}
+
+func TestIntervalSolver_EmptyInterval(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 100)
+	_, err := solver.Generate(rand.New())
+	if err != ErrUnsatisfiable {
+		t.Errorf("failed to generate point in empty interval: %v", err)
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:142]: This error message is wrong.

File: go/ct/gen/interval_solver_test.go [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+	}
+}
+
+func TestIntervalSolver_ExcludeEmptyIntervalHasNoEffect(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	clone := solver.Clone()
+	clone.Exclude(300, 250)
+	if !solver.Equals(clone) {
+		t.Fatalf("empty interval should not be modified by adding empty interval")
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:151]: this error message is wrong

File: go/ct/gen/interval_solver_test.go [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+	solver := NewIntervalSolver[int32](200, 400)
+	solver.Exclude(250, 300)
+
+	rnd := rand.New()
+	for i := 0; i < 100; i++ {
+		res, err := solver.Generate(rnd)
+		if err != nil {
+			t.Fatalf("error solving intervals: %v", err)
+		}
+		if res < 200 || res > 400 || (res >= 250 && res <= 300) {
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:165]: There should be a test case making sure that the boundaries are hit by the generation process.

File: go/ct/gen/interval_solver_test.go [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+		"lower":    func(s *IntervalSolver[int32]) { s.AddLowerBoundary(300) },
+		"upper":    func(s *IntervalSolver[int32]) { s.AddUpperBoundary(300) },
+		"equality": func(s *IntervalSolver[int32]) { s.AddEqualityConstraint(300) },
+	}
+
+	for name, setup := range tests {
+		t.Run(name, func(t *testing.T) {
+
+			solver := NewIntervalSolver[int32](200, 400)
+			solver.intervals = nil
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:194]: I would prefer creating the empty interval through public API.
```suggestion
			solver := NewIntervalSolver[int32](400, 200)
```

File: go/ct/gen/interval_solver_test.go [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+			check: func(v int64) bool { return v != math.MaxInt64 }},
+		"remove-max": {
+			setup: func(s *IntervalSolver[int64]) { s.Exclude(math.MaxInt64, math.MaxInt64) },
+			check: func(v int64) bool { return v != math.MaxInt64 }},
+		"remove-all-positive": {
+			setup: func(s *IntervalSolver[int64]) { s.Exclude(0, math.MaxInt64) },
+			check: func(v int64) bool { return v < 0 }},
+		"remove-all-negative": {
+			setup: func(s *IntervalSolver[int64]) { s.Exclude(math.MinInt64, 0) },
+			check: func(v int64) bool { return v >= 0 }},
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:303]: It should actually fail if 0 is generated.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

**Review by LuisPH3 — DISMISSED** [review]
:+1:

**Review by LuisPH3 — APPROVED** [review]
:+1:

File: go/ct/gen/interval_solver_test.go [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+func TestIntervalSolver_ExcludeEmptyIntervalHasNoEffect(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	clone := solver.Clone()
+	clone.Exclude(300, 250)
+	if !solver.Equals(clone) {
+		t.Errorf("excluding empty interval should have no effect")
+	}
+}
+
+func TestIntervalSolver_TestIntervals(t *testing.T) {
```
Comment (HerbertJordan) [inline go/ct/gen/interval_solver_test.go:181]:
```suggestion
func TestIntervalSolver_GenerateProducesValuesFromWithinTheIntervals(t *testing.T) {
```

**Review by HerbertJordan — APPROVED** [review]
Thanks for cleaning this up!

File: go/ct/gen/interval_solver.go
Comment (simonlechner) [inline go/ct/gen/interval_solver.go:1]: I think the general comment should also include whether the intervals are inclusive or exclusive.

**Review by simonlechner — APPROVED** [review]
Just a minor comment, looks good.

**Review by LuisPH3 — APPROVED** [review]
